indexing-service/src/main/java/org/apache/druid/guice/PeonProcessingModule.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.druid.collections.BlockingPool;
+import org.apache.druid.collections.DummyBlockingPool;
+import org.apache.druid.collections.DummyNonBlockingPool;
+import org.apache.druid.collections.NonBlockingPool;
+import org.apache.druid.guice.annotations.Global;
+import org.apache.druid.guice.annotations.Merging;
+import org.apache.druid.guice.annotations.Smile;
+import org.apache.druid.indexing.common.task.Task;
+import org.apache.druid.java.util.common.lifecycle.Lifecycle;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.query.DruidProcessingConfig;
+import org.apache.druid.query.ExecutorServiceMonitor;
+import org.apache.druid.query.NoopQueryProcessingPool;
+import org.apache.druid.query.QueryProcessingPool;
+import org.apache.druid.query.groupby.GroupByQueryConfig;
+import org.apache.druid.query.groupby.GroupByResourcesReservationPool;
+
+import java.nio.ByteBuffer;
+
+/**
+ * This module fulfills the dependency injection of query processing and caching resources: buffer pools and
+ * thread pools on Peon selectively. Only the peons for the tasks supporting queries need to allocate direct buffers
+ * and thread pools. Thus, this is separate from the {@link DruidProcessingModule} to separate the needs of the peons and
+ * the historicals
+ *
+ * @see DruidProcessingModule
+ */
+public class PeonProcessingModule implements Module
+{
+  private static final Logger log = new Logger(PeonProcessingModule.class);
+
+  @Override
+  public void configure(Binder binder)
+  {
+    DruidProcessingModule.registerConfigsAndMonitor(binder);
+  }
+
+  @Provides
+  @LazySingleton
+  public CachePopulator getCachePopulator(
+      @Smile ObjectMapper smileMapper,
+      CachePopulatorStats cachePopulatorStats,
+      CacheConfig cacheConfig
+  )
+  {
+    return DruidProcessingModule.createCachePopulator(smileMapper, cachePopulatorStats, cacheConfig);
+  }
+
+  @Provides
+  @ManageLifecycle
+  public QueryProcessingPool getProcessingExecutorPool(
+      Task task,
+      DruidProcessingConfig config,
+      ExecutorServiceMonitor executorServiceMonitor,
+      Lifecycle lifecycle
+  )
+  {
+    if (!task.supportsQueries()) {
+      if (config.isNumThreadsConfigured()) {
+        log.warn(
+            "Ignoring the configured numThreads[%d] because task[%s] of type[%s] does not support queries",
+            config.getNumThreads(),
+            task.getId(),
+            task.getType()
+        );
+      }
+      return NoopQueryProcessingPool.instance();
+    }
+    return DruidProcessingModule.createProcessingExecutorPool(config, executorServiceMonitor, lifecycle);
+  }
+
+  @Provides
+  @LazySingleton
+  @Global
+  public NonBlockingPool<ByteBuffer> getIntermediateResultsPool(Task task, DruidProcessingConfig config)
+  {
+    if (!task.supportsQueries()) {
+      return DummyNonBlockingPool.instance();
+    }
+    return DruidProcessingModule.createIntermediateResultsPool(config);
```

Comment:
Maybe invert the condition for readability:

```suggestion
    if (task.supportsQueries()) {
      return DruidProcessingModule.createIntermediateResultsPool(config);
    } else {
      return DummyNonBlockingPool.instance();
    }
```

Reply:
Same comment in other methods.